dttools/src/jx_match.c

```diff
@@ -0,0 +1,198 @@
+/*
+Copyright (C) 2016- The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file COPYING for details.
+*/
+
+#include <stdarg.h>
+#include <string.h>
+
+#include "jx.h"
+#include "jx_match.h"
+
+void *jx_match_null(struct jx *j) {
+	return jx_istype(j, JX_NULL) ? (void *) j : NULL;
+}
+
+int *jx_match_boolean(struct jx *j, int *v) {
+	if (jx_istype(j, JX_BOOLEAN)) {
+		if (v) {
+			*v = !!j->u.boolean_value;
+		}
+		return &j->u.boolean_value;
+	} else {
+		return NULL;
+	}
+}
+
+jx_int_t *jx_match_integer(struct jx *j, jx_int_t *v) {
+	if (jx_istype(j, JX_INTEGER)) {
+		if (v) {
+			*v = j->u.integer_value;
+		}
+		return &j->u.integer_value;
+	} else {
+		return NULL;
+	}
+}
```

Comment:
Returning a non-`const` pointer was actually intentional. Usually either

- the caller just wants to read the value, so put a copy in the `v` argument and check the return as a truth value.
- the caller wants to modify the value, so pass `NULL` in `v` and grab the returned pointer. The caller can do arbitrary reads/modifications directly on value in the structure without `malloc()`ing anything.

Are we making an assumption that JX structures are immutable? I don't think we have enough library built up to do that efficiently, and allowing direct modifications would be useful for incrementally building things.